### PR TITLE
Feature: added exclude-filter argument to keep certain image tags

### PR DESCRIPTION
**Pull request by angrox**
**Purpose of the PR**

Golang does not support the full capability of perl regex to filter image tags, e.g. to negate a regex. A new flag is added to exclude tags found by the --filter argument. This is useful when you want to keep certain images, e.g. all image tags which are currently in use by a K8s cluster.

Fixes #

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -117,7 +118,7 @@ Examples of filters
 | Untag tags that end with world                                                 | --filter `"<repository>:\w*world\b"`  |
 | Untag tags that include hello-world in their name                              | --filter `"<repository>:hello-world"` |
 | Untag all tags that are older than the duration                                | --filter `"<repository>:.*"`          |
-
+| Exclude all tags from untagging that end with -prod                            | --exclude-filter `".*-prod$"`          |
```
Comment (shizhMSFT) [inline README.md]: nit: `-prod`

File: README.md [hunk 2]
```diff
@@ -160,6 +161,19 @@ acr purge \
     --dry-run
 ```
 
+##### Exclude filter tag
+
+To exclude tags from untagging one or more --exclude-filter arguments can be added
```
Comment (shizhMSFT) [inline README.md]: nit: `--exclude-filter` for better representation.

File: cmd/acr/purge.go [hunk 3]
```diff
@@ -264,6 +270,21 @@ func getTagsToDelete(ctx context.Context,
 				// If a tag does not match the regex then it not added to the list no matter the LastUpdateTime
 				continue
 			}
+			isExcluded := false
+			for _, exFilter := range excludeFilter {
+				exFilterRegex, exFilterErr := regexp.Compile(exFilter)
```
Comment (shizhMSFT) [inline cmd/acr/purge.go]: `exFilter` is complied unnecessarily multiple times. You should pre-compile it like `filter`.

File: cmd/acr/purge.go [hunk 4]
```diff
@@ -240,6 +245,7 @@ func getTagsToDelete(ctx context.Context,
 	acrClient api.AcrCLIClientInterface,
 	repoName string,
 	filter *regexp.Regexp,
+	excludeFilter []string,
```
Comment (shizhMSFT) [inline cmd/acr/purge.go]: You should pass a `*regexp.Regexp` instead of `[]string` for `excludeFilter`. The reason is explained in the next comment.

File: cmd/acr/purge.go [hunk 5]
```diff
@@ -264,6 +270,21 @@ func getTagsToDelete(ctx context.Context,
...
+				exFilterRegex, exFilterErr := regexp.Compile(exFilter)
+				if exFilterErr != nil {
+					return nil, "", exFilterErr
+				}
+				if exFilterRegex.MatchString(*tag.Name) {
+					isExcluded = true
+					continue
+				}
+			}
+			if isExcluded {
```
Comment (shizhMSFT) [inline cmd/acr/purge.go]: Once the `excludeFilter` is a `*regexp.Regexp`, the variable `isExcluded` can be simplified.

File: cmd/acr/purge_test.go [hunk 6]
```diff
@@ -26,7 +26,7 @@ func TestPurgeTags(t *testing.T) {
 		assert := assert.New(t)
 		mockClient := &mocks.AcrCLIClientInterface{}
 		mockClient.On("GetAcrTags", testCtx, testRepo, "", "").Return(notFoundTagResponse, errors.New("testRepo not found")).Once()
-		deletedTags, err := purgeTags(testCtx, mockClient, testLoginURL, testRepo, "1d", "[\\s\\S]*")
+		deletedTags, err := purgeTags(testCtx, mockClient, testLoginURL, testRepo, "1d", "[\\s\\S]*", nil)
```
Comment (shizhMSFT) [inline cmd/acr/purge_test.go]: You need real test cases instead of simply passing `nil` to it.

**Comment by fiv21** [comment]
How can we help to get this implemented in the short term? 🤔

**Comment by sashokbg** [comment]
Hello @northtyphoon , why was this MR closed ? Can we finish the work (I can try if needed) :)
